task/packrat/src/ereport.rs

```diff
@@ -0,0 +1,405 @@
+    },
+    Flushed {
+        committed_ena: u64,
+        flushed: usize,
+    },
+    RestartIdMismatch {
+        current_restart_id: u128,
+    },
+    MetadataError(#[count(children)] MetadataError),
+    MetadataEncoded {
+        len: u32,
+    },
+    EreportError(#[count(children)] EreportError),
+    Reported {
+        start_ena: u64,
+        reports: u8,
+        limit: u8,
+    },
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, counters::Count)]
+enum MetadataError {
+    TooLong,
+    PartNumberNotUtf8,
+    SerialNumberNotUtf8,
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, counters::Count)]
+enum EreportError {
+    TaskIdOutOfRange,
+    TooLong,
+}
+
+counted_ringbuf!(Trace, 16, Trace::None);
+
+impl EreportStore {
+    pub(crate) fn new(
+        EreportBufs {
+            ref mut storage,
+            ref mut recv,
+        }: &'static mut EreportBufs,
+    ) -> Self {
+        let now = sys_get_timer().now;
+        storage.initialize(config::TASK_ID, now);
+        let image_id = {
+            let id = kipc::read_image_id();
+            u64::to_le_bytes(id)
+        };
+
+        Self {
+            storage,
+            recv,
+            image_id,
+            restart_id: None,
+        }
+    }
+}
+
+impl EreportStore {
+    pub(crate) fn deliver_ereport(
+        &mut self,
+        msg: &RecvMessage,
+        data: LenLimit<Leased<idol_runtime::R, [u8]>, RECV_BUF_SIZE>,
+    ) -> Result<(), RequestError<Infallible>> {
+        data.read_range(0..data.len(), self.recv)
+            .map_err(|_| ClientError::WentAway.fail())?;
+        let timestamp = sys_get_timer().now;
+        let result = self.storage.insert(
+            msg.sender.0,
+            timestamp,
+            &self.recv[..data.len()],
+        );
+        ringbuf_entry!(Trace::EreportReceived {
+            src: msg.sender,
+            len: data.len() as u32,
+            result,
+        });
+        Ok(())
+    }
+
+    pub(crate) fn read_ereports(
+        &mut self,
+        request_id: ereport_messages::RequestIdV0,
+        restart_id: ereport_messages::RestartId,
+        begin_ena: ereport_messages::Ena,
+        limit: u8,
+        committed_ena: ereport_messages::Ena,
+        data: Leased<idol_runtime::W, [u8]>,
+        vpd: Option<&VpdIdentity>,
+    ) -> Result<usize, RequestError<EreportReadError>> {
+        ringbuf_entry!(Trace::ReadRequest {
+            restart_id: restart_id.into()
+        });
+
+        /// Byte indicating the end of an indeterminate-length CBOR array or
+        /// map.
+        const CBOR_BREAK: u8 = 0xff;
+
+        let current_restart_id =
+            self.restart_id.ok_or(EreportReadError::RestartIdNotSet)?;
+        // Skip over a header-sized initial chunk.
+        let first_data_byte = size_of::<ereport_messages::ResponseHeader>();
+
+        let mut position = first_data_byte;
+        let mut first_written_ena = None;
+
+        // Begin metadata map.
+        data.write_at(position, 0xbf)
```

Comment:
Should we also define this as a constant (up near `CBOR_BREAK`)?

Reply:
we use the `CBOR_BREAK` byte twice, but we only use the "begin map" and "begin list" bytes once, so it felt less important...i'll change them for consistency anyway, since i agree it reads nicer.